### ISSUE TITLE
Spell out where views are allowed

### ISSUE THREAD

**Issue by crusaderky** [data-apis/array-api-compat#331]
In https://data-apis.org/array-api/latest/design_topics/copies_views_and_mutation.html, the Standard says

> Array API consumers are strongly advised to avoid any mutating operations when an array object may [...] be a “view” [...] It is not always clear, however, when a library will return a view and when it will return a copy. This standard does not attempt to specify this—libraries may do either.

The above is fine after `__getitem__` , `asarray(..., copy=None)`, `astype(..., copy=False)`, and similar functions that are explicitly explained by the standard to potentially return views.

However, there are a few corner cases where views _could_ be possible but a normal user is very unlikely to think about them.
I just stumbled on one in https://github.com/data-apis/array-api-compat/pull/298, where `array_api_compat.torch.sum(x, dtype=x.dtype, axis=())` was accidentally returning `x` instead of a copy of it.

There are a few more cases where a library could try to be smart; for example

- search functions (`min`, `max`, other?) could return a view to the minimum/maximum point
- replacement functions (`minimum`, `maximum`, `clip`, `where`) could return one of the input arrays when there is nothing to do
- same for arithmetic functions (`__add__` / `__sub__` vs. 0, `__mul__` / `__div__` vs. 1, etc.)
- same for sort functions when they realise the input is already sorted
- possibly more

In real life, I expect end users to assume that the above functions will always return a copy.
I think the standard should spell this out, limiting the possibily of views to an explicit list of allowed functions:
- `__getitem__`
- `asarray`
- `astype`
- `__dlpack__`
- `from_dlpack`
- `reshape`
- `broacast_to`
- `broadcast_arrays`
- ...more?

**Comment by ev-br** [comment]
https://github.com/data-apis/array-api-compat/pull/333 loops over unary functions and fixes `trunc`, `ceil` and `floor`. Potentially there are more potentially trickier cases, but those are not tested for in gh-333 just yet.

**Comment by ev-br** [comment]
To clarify my position on the matter: Since the spec does not mandate whether any given function returns a view or a copy, this is "unspecified, thus implementation-defined". This is one place where the whole Array API abstraction leaks (inevitably, IMO).

Thus, what's left in the array-api-compat level is to decide what we do about it. The only reasonable thing IMO is to declare that for a bare array library X, its wrapped version behaves the same. A small favor to users is to extend this to library versions: if a library version X.Y returns a view, so does its wrapped version.
If we all agree to this, maybe it's worth spelling it out explicitly somewhere in the docs. Not sure where though.

This of course does not shield users from surprises where, say, numpy returns a view but jax.numpy returns a copy. Not much we can do about it, maybe we could document it somewhere (not sure where either).
